### PR TITLE
[Merged by Bors] - Feat: Port data.prod.tprod

### PR DESCRIPTION
**Pull request by arienmalec**
Port of data.prod.tprod

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Prod/TProd.lean [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+! leanprover-community/mathlib commit 7b78d1776212a91ecc94cf601f83bdcc46b04213
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.List.Nodup
+/-!
+# Finite products of types
+
+This file defines the product of types over a list. For `l : List ι` and `α : ι → Type*` we define
+`list.TProd α l = l.foldr (λ i β, α i × β) PUnit`.
```
Comment (dupuisf) [inline Mathlib/Data/Prod/TProd.lean:16]:
```suggestion
This file defines the product of types over a list. For `l : List ι` and `α : ι → Type _` we define
`List.TProd α l = l.foldr (λ i β, α i × β) PUnit`.
```

File: Mathlib/Data/Prod/TProd.lean [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+The application of this type is finitary product measures, but it could be used in any
+construction/theorem that is easier to define/prove on binary products than on finitary products.
+
+* Once we have the construction on binary products (like binary product measures in
+  `MeasureTheory.prod`), we can easily define a finitary version on the type `TProd l α`
+  by iterating. Properties can also be easily extended from the binary case to the finitary case
+  by iterating.
+* Then we can use the equivalence `List.TProd.piEquivTProd` below (or enhanced versions of it,
+  like a `MeasurableEquiv` for product measures) to get the construction on `∀ i : ι, α i`, at
+  least when assuming `[Fintype ι] [Encodable ι]` (using `Encodable.sorted_univ`).
```
Comment (dupuisf) [inline Mathlib/Data/Prod/TProd.lean:29]:
```suggestion
  least when assuming `[Fintype ι] [Encodable ι]` (using `Encodable.sortedUniv`).
```

File: Mathlib/Data/Prod/TProd.lean [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+end List
+
+namespace Set
+
+open List
+
+/-- A product of sets in `TProd α l`. -/
+@[simp]
+protected def TProd : ∀ (l : List ι) (_t : ∀ i, Set (α i)), Set (TProd α l)
```
Comment (dupuisf) [inline Mathlib/Data/Prod/TProd.lean:156]:
```suggestion
protected def tprod : ∀ (l : List ι) (_t : ∀ i, Set (α i)), Set (TProd α l)
```
I guess it should be either this or `tProd`.

**Review by dupuisf — COMMENTED** [review]
Just a few name changes, otherwise LGTM

bors d+

File: Mathlib/Data/Prod/TProd.lean (hunk 1 above)
Comment (arienmalec) [reply Mathlib/Data/Prod/TProd.lean]: addressed

**Pull request closed without merging**